File: leetcode/576. Out of Boundary Paths/main.dart
class Solution {
  int findPaths(int m, int n, int N, int x, int y) {
    final int M = 1000000000 + 7;

    List<List<int>> dp = initDP(m, n);

    dp[x][y] = 1;

    int count = 0;

    for (int moves = 1; moves <= N; moves++) {
      List<List<int>> temp = initDP(m, n);

      for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
          if (i == m - 1) count = (count + dp[i][j]) % M;
          if (j == n - 1) count = (count + dp[i][j]) % M;
          if (i == 0) count = (count + dp[i][j]) % M;
          if (j == 0) count = (count + dp[i][j]) % M;

          temp[i][j] =
              (((i > 0 ? dp[i - 1][j] : 0) + (i < m - 1 ? dp[i + 1][j] : 0)) %
                          M +
                      ((j > 0 ? dp[i][j - 1] : 0) +
                              (j < n - 1 ? dp[i][j + 1] : 0)) %
                          M) %
                  M;
        }
      }

      dp = temp;
    }

    return count;
  }

  List<List<int>> initDP(int rows, int cols) {
    return List.generate(rows, (i) => List<int>.filled(cols, 0));
  }
}

File: leetcode/576. Out of Boundary Paths/main.go
package main

const Modulo = 1e9 + 7

func findPaths(m int, n int, maxMove int, startRow int, startColumn int) int {
	dp := initGridDP(m, n)

	dp[startRow][startColumn] = 1
	count := 0

	for move := 1; move <= maxMove; move++ {
		newDP := initGridDP(m, n)
		for i := 0; i < m; i++ {
			for j := 0; j < n; j++ {
				if i == 0 {
					count = (count + dp[i][j]) % Modulo
				}
				if j == 0 {
					count = (count + dp[i][j]) % Modulo
				}
				if i == m-1 {
					count = (count + dp[i][j]) % Modulo
				}
				if j == n-1 {
					count = (count + dp[i][j]) % Modulo
				}

				if i > 0 {
					newDP[i][j] += dp[i-1][j]
				}
				if i < m-1 {
					newDP[i][j] += dp[i+1][j]
				}

				newDP[i][j] = newDP[i][j] % Modulo

				if j > 0 {
					newDP[i][j] += dp[i][j-1]
				}
				if j < n-1 {
					newDP[i][j] += dp[i][j+1]
				}

				newDP[i][j] = newDP[i][j] % Modulo
			}
		}

		dp = newDP
	}

	return count
}

func initGridDP(rows, cols int) [][]int {
	dp := make([][]int, rows)
	for i := range dp {
		dp[i] = make([]int, cols)
	}

	return dp
}
